reccmp/isledecomp/compare/report.py

```diff
@@ -0,0 +1,185 @@
+from datetime import datetime
+from dataclasses import dataclass
+from typing import Literal, Iterable, Iterator
+from pydantic import BaseModel, ValidationError
+from pydantic_core import from_json
+from .diff import CombinedDiffOutput
+
+
+class ReccmpReportDeserializeError(Exception):
+    """The given file is not a serialized reccmp report file"""
+
+
+@dataclass
+class ReccmpComparedEntity:
+    orig_addr: str
+    name: str
+    accuracy: float
+    recomp_addr: str | None = None
+    is_effective_match: bool = False
+    is_stub: bool = False
+    diff: CombinedDiffOutput | None = None
+
+
+class ReccmpStatusReport:
+    # The filename of the original binary.
+    # This is here to avoid comparing reports derived from different files.
+    # TODO: in the future, we may want to use the hash instead
+    filename: str
+
+    # Creation date of the report file.
+    timestamp: datetime
+
+    # Using orig addr as the key.
+    entities: dict[str, ReccmpComparedEntity]
+
+    def __init__(self, filename: str, timestamp: datetime | None = None) -> None:
+        self.filename = filename
+        if timestamp is not None:
+            self.timestamp = timestamp
+        else:
+            self.timestamp = datetime.now().replace(microsecond=0)
+
+        self.entities = {}
+
+
+def _get_entity_for_addr(
+    samples: Iterable[ReccmpStatusReport], addr: str
+) -> Iterator[ReccmpComparedEntity]:
+    """Helper to return entities from xreports that have the given address."""
+    for sample in samples:
+        if addr in sample.entities:
+            yield sample.entities[addr]
+
+
+def _accuracy_sort_key(entity: ReccmpComparedEntity) -> float:
+    """Helper to sort entity samples by accuracy score.
+    100% match is preferred over effective match.
+    Effective match is preferred over any accuracy.
+    Stubs rank lower than any accuracy score."""
+    if entity.is_stub:
+        return -1.0
+
+    if entity.accuracy == 1.0:
+        if not entity.is_effective_match:
+            return 1000.0
+
+    if entity.is_effective_match:
+        return 1.0
+
+    return entity.accuracy
+
+
+def combine_reports(samples: list[ReccmpStatusReport]) -> ReccmpStatusReport:
+    """Combines the sample reports into a single report.
+    The current strategy is to use the entity with the highest
+    accuracy score from any report."""
+    assert len(samples) > 0
+
+    # TODO: hack
+    assert all(samples[0].filename == s.filename for s in samples)
```

Comment:
Still planning on addressing this before merging? If it's good enough, you can probably replace the comment by a non-TODO